### ISSUE TITLE
[Bug]: PC0038 - do not trigger when its a boolean and you use exit

### ISSUE THREAD

**Issue by guidorobben** [ALCops/Analyzers#430]
### Analyzer

PlatformCop

### Issue type

False positive (diagnostic fires when it shouldn't)

### Rule ID

PC0038

### Description

If I have a function that returns a boolean, currently we need to specify the false and the true. But most of the times I just use exit instead of exit(false). In my opinion, the extra (false) is not needed

### AL code to reproduce

```al
 procedure GetContainer(var Item: Record Item): Boolean
    begin
        if Item."No." = '' then
            exit; //This is now triggered and want me to change to exit(false);

        exit(true);
    end;
```

**Comment by guidorobben** [comment]
This is also true for integers and decimals.

**Comment by lv-janpieter** [comment]
I think the current behavior is intentional. The analyzer cannot determine whether the missing return value is deliberate or whether exit(false); (or another explicit value) was intended. The warning encourages developers to make that choice explicit.

In my experience, this kind of check prevents subtle bugs that can otherwise go unnoticed. Several languages even require all code paths to return an explicit value before the code will compile.

An explicit exit(...) or assignment to the return variable makes the intent clear and resolves the warning.

For that reason, I would vote to keep the current behavior. In my opinion, suppressing this warning would reduce the effectiveness of the analyzer and make it easier for unintended return values to slip into production code.

**Comment by Arthurvdv** [comment]
Thanks for the report and the discussion, both viewpoints are reasonable. Before deciding, we investigated two things: how other languages and linters handle this exact situation, and how often the pattern actually occurs in Microsoft's own code. Sharing the results here, along with a recommendation.

## First, a reframing

Strictly speaking this is not a false positive: PC0038 fires exactly as designed. A bare `exit;` in a value-returning procedure is *deterministic, documented* AL behavior (it returns the type's default: `false`, `0`, `''`). So the question is not "is the diagnostic wrong?" but "should the rule treat an implicit default return as acceptable?" That's a policy decision, and worth deciding on evidence.

One note on the type angle ("this is also true for integers and decimals"): in AL *every* type has a well-defined default, not just the basic ones (empty record, null guid, `0DT`, ...). So "the type has a default" cannot be the criterion for a carve-out, that would apply to everything and would gut the rule. The only defensible carve-out argument is *idiom*: bare `exit;` as "false / not found" in Boolean guard clauses mirrors the platform's own `Get`/`FindFirst` semantics.

## How other languages handle this

| Language / tool | Rule | Severity | On by default? | Type carve-out? |
|---|---|---|---|---|
| C# compiler | CS0161 | **Error** | Always (can't disable) | No |
| Java / Rust / Swift | (compiler) | **Error** | Always | No |
| TypeScript | `noImplicitReturns` | Error when enabled | **No** — opt-in, not even in `strict` | No |
| ESLint | `consistent-return` | Configurable | **No** — not in `recommended` | No |
| Pylint | R1710 `inconsistent-return-statements` | Refactor (style) | **Yes** | No |
| Ruff | RET502/RET503 | Lint | **No** — opt-in | No |
| Go | naked `return` with named results | Legal; style guidance only | n/a | No |
| Delphi (closest AL relative) | W1035 "Return value might be undefined" | **Warning** | **Yes** | No |
| PHPStan | missing return path | Error at level ≥ 4 | **No** — default level 0 | No |

Three patterns stand out:

1. **Languages with no default value make it a hard error** (C#, Java, Rust). **Languages with a deterministic default make it at most a warning, often opt-in** (TypeScript, Python, Go, PHP). AL is firmly in the second group.
2. **Delphi is the most instructive precedent.** Like AL, it has a zero-initialized implicit `Result` variable, so the returned value is always well-defined. It *still* warns by default (W1035), on exactly the reasoning @lv-janpieter gave: the compiler cannot know whether the default was intended or forgotten, so it asks the developer to make intent explicit.
3. **No tool in any surveyed ecosystem implements a type-specific carve-out** ("allow bare return for bool but not int"). There is no precedent for option (b) below.

Also worth noting: ESLint's `consistent-return` deliberately only flags *mixed* styles (some paths return a value, some don't) — which is precisely the reported code shape (`exit;` + `exit(true)` in one procedure), and precisely what our own LC0097 covers for named returns.

## How often this occurs in practice (BaseApp w1-28.3, 8,094 files)

We compiled the Base Application with PC0038 enabled:

- **1,465 hits** out of 13,755 value-returning procedures (**10.7%**)
- By return type: Boolean 42.7%, Text 19.8%, Decimal 9.8%, Integer 8.5%, Code 6.9%, rest <8%
- By shape:
  - **62.7%** — fallthrough on some branch while other branches `exit(value)` (no bare `exit;` involved at all)
  - **21.4%** — pure fallthrough (no exit anywhere)
  - **12.9%** — mixed bare `exit;` + `exit(value)` (the shape reported here)
  - **3.0%** — only bare `exit;`
- The exact reported scenario — Boolean with a bare `exit;` guard clause — is **~7%** of all hits.

Two takeaways:

1. The bare-`exit;` idiom is real but it is a **minority** of what PC0038 finds (~16% of hits contain a bare `exit;` at all). Suppressing it keeps ~84% of the rule's findings intact — so a carve-out would *not* gut the rule.
2. Conversely, the dominant finding (62.7%) is a branch that silently falls through while sibling branches return explicit values. That's the genuinely bug-prone shape the rule exists for, and it's unaffected by this discussion.

## Options

**(a) Keep as-is.** Matches the Delphi precedent, keeps intent explicit. PC0038 is already only **Info** severity, deliberately low-ceremony: it doesn't fail builds and is trivially disabled per-project via ruleset. Cost: 10.7% of BaseApp-style procedures flagged, and friction for teams where `exit;` = `exit(false)` is a house idiom.

**(b) Never flag bare `exit;`.** Removes the friction, but creates an inconsistency: `if X then exit;` would pass while an equivalent `if X then begin end;` fallthrough fails, even though both return the default. It also hides the genuinely suspicious *mixed* case for unnamed returns. No ecosystem precedent treats bare-return-with-default as universally fine while flagging fallthrough.

**(c) Boolean-only carve-out.** No precedent in any surveyed tool, and the justification ("has a default") applies to every AL type, so the boundary is arbitrary and will immediately generate follow-up requests for Integer/Decimal/Text (as already happened in this thread).

## Recommendation

Keep the current behavior — the Delphi precedent, the severity (Info), and the data (84% of findings are unrelated to bare `exit;`) all support it. The rule is deliberately low-ceremony: it doesn't fail builds, and teams for whom `exit;` = `exit(false)` is a house idiom can disable PC0038 per-project via ruleset. Explicitly recommend **against** (b) and (c).

@guidorobben, our analysis points to keeping the rule as-is, but we're open to discussion — does the reasoning above work for you, or do you see a case we've missed? If you're on board, we'll close this as by-design.

---
*Generated with Claude Fable 5*

**Comment by guidorobben** [comment]
Thanks for the explanation. I will throw this in the  development group and we will decide what todo with the rule. But looking at the data, I say keep the rule as is. So it will be a decision for us to decide if we turn off the rule or not.  

So the issue can be closed.

**Comment by Arthurvdv** [comment]
Thanks for circling back @guidorobben on this and glad the data helped to shape the decision.

If you're open to sharing what the development group landed on with the reasoning behind it, I'de love to hear that feedback. This could help other teams who might stumble on this issue while evaluating the same trade-off for their codebase. 

Absolutely no obligation though, I completely understand these are internal discussions.

**Comment by guidorobben** [comment]
At the moment we decided to turrn the rule off. The reason it is producing a lot of warnings that clutter the results too much now. It is a lot of work to fix the warnings and we will do it in our own schedule. 

And also we still haven't decided for 100%, is not setting the result value is a bad thing.

**Comment by lv-janpieter** [comment]
That's the effort that comes with adopting any new analyzer.

You already have an existing codebase that probably works reasonably well, or at least no one has noticed the existing issues yet. If you want analyzers to help improve newly written code, you first need to decide how to handle the existing violations. That usually means either fixing them gradually or suppressing them explicitly.

I generally encourage developers to fix new analyzer warnings whenever they touch an extension. In many cases the effort is quite limited, especially with modern AI tools such as Copilot or Cursor helping with the repetitive changes. I don't expect a devloper to fix hundreds or thousands of existing warnings overnight. However, when touching a piece of code, fixing the warnings in that area is usually a worthwhile investment. [(Boy Scout Rule)](https://deviq.com/principles/boy-scout-rule/)

What is often overlooked is the cost of bugs that make it into released software. Someone has to discover them, create an issue, a developer has to investigate and fix them, deploy the solution, and then tested again. Preventing those issues early is usually significantly cheaper than dealing with them later.
